frontend/expr_inferrer.py

```diff
@@ -331,80 +329,119 @@ def infer_dict_comprehension(node, context):
         The iteration target should be always a variable name.
     """
     local_context = Context(parent_context=context)
-    infer_generators(node.generators, local_context, node.lineno)
-    key_type = infer(node.key, local_context)
-    val_type = infer(node.value, local_context)
-    return z3_types.Dict(key_type, val_type)
+    infer_generators(node.generators, local_context, node.lineno, solver)
+    key_type = infer(node.key, local_context, solver)
+    val_type = infer(node.value, local_context, solver)
+    return solver.z3_types.dict(key_type, val_type, solver)
 
 
-def _get_args_types(args, context):
+def _get_args_types(args, context, instance, solver):
     """Return inferred types for function call arguments"""
     # TODO kwargs
 
-    args_types = ()
+    args_types = () if instance is None else (instance,)
     for arg in args:
-        args_types = args_types + (infer(arg, context),)
+        args_types = args_types + (infer(arg, context, solver),)
     return args_types
 
 
-def infer_func_call(node, context):
+def infer_func_call(node, context, solver):
     """Infer the type of a function call, and unify the call types with the function parameters"""
-    func_type = infer(node.func, context)
-    args_types = _get_args_types(node.args, context)
+    instance = None
+    if isinstance(node.func, ast.Attribute):
+        called, instance = infer(node.func, context, solver, True)
+    else:
+        called = infer(node.func, context, solver)
+    args_types = _get_args_types(node.args, context, instance, solver)
 
-    result_type = z3_types.new_z3_const("call")
+    result_type = solver.new_z3_const("call")
 
     # TODO covariant and invariant subtyping
-    z3_types.type_solver.add(func_type == z3_types.Funcs[len(args_types)](args_types + (result_type,)),
-                             fail_message="Call in line {}".format(node.lineno))
+
+    solver.add(axioms.call(called, args_types, result_type, solver.z3_types),
+               fail_message="Call in line {}".format(node.lineno))
+
+    return result_type
+
+
+def infer_attribute(node, context, from_call, solver):
+    """Infer the type of attribute access
+    
+    Cases:
+        - Instance attribute access. ex:
```

Comment:
Cool. Does that mean we can also handle attribute accesses of class fields on the class fields?